### ISSUE TITLE
Visual Studio 2008 Support and Python 2.7

### ISSUE THREAD

**Issue by patricksnape** [opencv/opencv#8481]
Python 2.7 compilation on Windows requires compiling with Visual Studio 2008 (VS2008). This is highly unfortunate, but is the reality. Is the OpenCV team still supporting this? If not, do you officially drop your support for the python.org distributed version of Python 2.7? If you do still support it, would you be willing to support me in setting up a VS2008 build environment for OpenCV?

For some context of this issue, I just spent a lot of time getting OpenCV 3.2.0 to build for [conda-forge](https://github.com/conda-forge/opencv-feedstock/pull/50), which is widely used in the Python community for installing scientific Python packages. I field a lot of responses about installing and building OpenCV and would prefer not to have to re-patch OpenCV after every release due to C++11 features creeping in. 

I also have a number of patches [here](https://github.com/conda-forge/opencv-feedstock/pull/50) that I will bring upstream if you are willing that bring back the ability to compile OpenCV for Python 2.7 on Windows.

cc @jakirkham

**Comment by alalek** [comment]
Do you observe any issues with OpenCV Python 2.7 module compiled by other versions of MSVS?

**Comment by patricksnape** [comment]
Python 2.7 has a hard requirement for VS 2008 for compatibility reasons [unfortunately](https://mail.python.org/pipermail/python-dev/2012-August/121412.html). See [Steve Dower's excellent post on the topic](http://stevedower.id.au/blog/building-for-python-3-5/). So whether it builds with other versions of MSVC is not relevant to this discussion (as it pertains to providing automated, fully compatible builds of Python 2.7 packages).

**Comment by jakirkham** [comment]
The other problem to note is even if someone is able to build Python 2.7 with a different MSVC version, which will be a pretty serious patching effort and some do this, one runs afoul of a new problem. They can't use any wheels built for Windows and Python 2.7 on PyPI. So they can no long pip install a precompiled NumPy or SciPy for instance. Instead they will have to build everything from source with the same MSVC that they built there interpreter with. This requires patching of `distutils` to build things against this non-standard compiler, which is pretty painful.

Reflecting on this, I think it is clear why there are very few people that use a non-standard MSVC. After all one must have exceptional knowledge of compilers, Windows, CPython, and MSVC to even attempt this let alone succeed. Even the few people I know who have accomplished this are trying desperately to get rid of that maintenance burden, which basically means dropping Python 2.7.

So it really isn't reasonable to even "support" Python 2.7 in this non-standard way for a larger group of people. One really has to either build and test against VS 2008 as @patricksnape has nicely laid out or simply drop Python 2.7 support.

**Comment by jakirkham** [comment]
Any news on this issue? TBH whether the decision is to drop VS 2008/Python 2.7 support or keep and test VS 2008/Python 2.7, it will be a marked improvement over the quasi-supported VS2008/Python 2.7 situation we have now. To clarify, not having to patch OpenCV substantially each release to bring it into compliance would be the improvement.

**Comment by alalek** [comment]
Some comments from me:
- We are able to [build](http://pullrequest.opencv.org/buildbot/builders/master_noOCL-win64-vc14/builds/10069) Python 2.7 extension (module) with MSVS2015 and we don't observe critical conflicts of different MSVC runtimes.
- OpenCV's build system is based on CMake. Python `distutils` is not used and there is no plan to support it.
- Mentioned MSVC 2008 for Python is not supported by Microsoft.

**Comment by jakirkham** [comment]
> We are able to build Python 2.7 extension (module) with MSVS2015...

I'm not questioning whether one *can* build it with a different runtime. Have even noted people above that actually *do* this. So the idea is not foreign to me. However after suggesting this to conda-forge last year with use cases like this in mind, the [decision]( https://github.com/conda-forge/conda-forge.github.io/issues/64 ) was not to support it. While it might be feasible there were enough people against it or unwilling to help, that the maintenance burden could not be reasonably handled by those interested nor the resources dedicated justified.

> OpenCV's build system is based on CMake. Python distutils is not used and there is no plan to support it.

I think something is getting lost in translation. We are not asking that OpenCV start building with `distutils`. We are asking that OpenCV be built and tested in a way that is compatible with `pip` or `conda` installable binary versions of things like NumPy, SciPy, etc. for Python 2.7. These will all be built with MSVC 2008 as well. There are no plans on our end to build them with anything else (note the linked conda-forge discussion above). That is assuming OpenCV plans to keep Windows Python 2.7 as a supported target.

> ...and we don't observe critical conflicts of different MSVC runtimes.
> Mentioned MSVC 2008 for Python is not supported by Microsoft.

The thing is other MSVCs are not supported for Python 2.7. Even Microsoft's own Steve Dower [concluded]( https://mail.python.org/pipermail/python-dev/2013-November/130599.html ) after looking at getting MSVC 2010 support for Python 2.7 that this was not reasonable to support and actually quite dangerous. To quote Steve Dower, "the best solution is always going to be for users to install VS 2008, or at least VC9...". After that Microsoft put together the [Microsoft Visual C++ Compiler for Python 2.7]( https://www.microsoft.com/en-us/download/details.aspx?id=44266 ), which was intended to simplify the work of developers trying to target MSVC 2008.

To summarize, while I get that MSVC 2008 is quite old, the reality is so is Python 2.7. To truly maintain compatibility with Python 2.7 on Windows, one must be willing to maintain compatibility with the tools and libraries used to build and operate it. If that is not something that OpenCV devs want to do, I completely get that. However, I think it is time to make that decision and be clear about it knowing everything we have shared.

**Comment by patricksnape** [comment]
If the official position from OpenCV is that they don't want to support MSVC 2008 then we just want that clarified. We will then cease to build Python 2.7 OpenCV builds for Windows and direct any complaints to the appropriate wiki (or this issue for example).

**Comment by mshabunin** [comment]
Sorry for delay. We discussed this topic several times and decided not to support VS 2008, main reasons are:
- good support needs resources - more complex CI configuration and more human attention; we had an idea to run weekly or pre-release builds and occasionally fix all issues, but this approach would require even more effort than pre-commit builds.
- we are thinking about enabling C++11 by default in both implementation and interface. When this happen support of old compilers will be almost impossible.

**Comment by vpisarev** [comment]
Generally, @mshabunin said it all. I shall add that trying to support older compilers in OpenCV is completely unfeasible task. We (opencv team) are trying to be very conservative, but C++11 is 6 years old already, VS 2008 is 9 years old and Python 3.x is almost 9 years old. So it's time to move on. OpenCV + Python users have at least 2 options: 1) use Python 3.x, 2) build OpenCV python module completely statically with VS2015 (or later) using OpenCV build system. In the case of static build cv2.pyd will only have dependency of python27.dll and Windows system DLLs, there will be no dependency of MSVCR*.dll or anything. So this is non-issue in general.

**Comment by patricksnape** [comment]
@vpisarev @mshabunin Thanks a lot for your response - I really appreciate you taking the time to discuss this. This is great news for me personally as I no longer have to maintain patches for Python 2.7 and 3.4 😆
